### PR TITLE
pub use std::simd::StdFloat;

### PR DESCRIPTION
**Pull request by workingjubilee**
Syncs portable-simd up to commit rust-lang/portable-simd@03f6fbb21e6050da2a05b3ce8f480c020b384916,
Diff: https://github.com/rust-lang/portable-simd/compare/533f0fc81ab9ba097779fcd27c8f9ea12261fef5...03f6fbb21e6050da2a05b3ce8f480c020b384916

This sync requires a little bit more legwork because it also introduces a trait into `std::simd`, so that it is no longer simply a reexport of `core::simd`. Out of simple-minded consistency and to allow more options, I replicated the pattern for the way `core::simd` is integrated in the first place, however this is not necessary if it doesn't acquire any interdependencies inside `std`: it could be a simple crate reexport. I just don't know yet if that will happen or not.

To summarize other misc changes:
- Shifts no longer panic, now wrap on too-large shifts (like `Simd` integers usually do!)
- mask16x32 will now be many i16s, not many i32s... 🙃
- `#[must_use]` is spread around generously
- Adjusts division, float min/max, and `Mask::{from,to}_array` internally to be faster
- Adds the much-requested `Simd::cast::<U>` function (equivalent to `simd.to_array().map(|lane| lane as U)`)

### CODE REVIEW

**Comment by workingjubilee** [comment]
Apologies if you had already started review, I pulled in the latest commits from portable-simd. I don't plan to move the sync point ahead again, it's just that we received so many requests for this `Simd::cast` function since we first merged into nightly, and we only recently cleared the other barriers for it, so I wanted to have it included as soon as you're ready.

**Comment by Mark-Simulacrum** [comment]
Just to confirm: these changes are seeing upstream review, right? Typically the subtree part of a subtree bump doesn't need (detailed) review, since it's already been reviewed.

The std changes here look OK to me from an integration perspective -- r=me after confirming that the other changes have been reviewed upstream (and you're not desiring another pass or specific examination of any of them)

It may make sense to just glob re-export some part of std_simd rather than specifically exporting the StdFloat trait, but there's advantages to this approach to and it seems OK for now.

**Comment by workingjubilee** [comment]
'kayo! I will make a note of that in the future!

And yes indeed, we are making sure we review each other's work here. I only really wanted to make sure the way I pushed things into `std` this time around seemed okay.
@bors r=Mark-Simulacrum
